api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,169 @@
+	}
+
+	async fetchUserID(identifier: string) {
+		const user = await this.knex
+			.select('id')
+			.from('directus_users')
+			.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
+			.first();
+
+		return user?.id;
+	}
+
+	async getUserID(payload: Record<string, any>) {
+		const { provider, emailKey, identifierKey, nameKey, givenNameKey, familyNameKey, allowPublicRegistration } =
+			this.config;
+
+		const email = payload[emailKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'];
+		const identifier = payload[identifierKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'];
+		const userID = await this.fetchUserID(identifier);
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[SAML] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		const firstName = payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'];
+		const lastName = payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'];
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: firstName,
+				last_name: lastName,
+				email: email,
+				external_identifier: identifier.toLowerCase(),
+				role: this.config.defaultRoleId,
+			});
+			const userID = await this.fetchUserID(identifier);
+			if (userID) return userID;
+		} catch (e) {
+			if (e instanceof RecordNotUniqueException) {
+				logger.warn(e, '[SAML] Failed to register user. User not unique');
+				throw new InvalidProviderException();
+			}
+			throw e;
+		}
+
+		throw new InvalidCredentialsException();
+	}
+
+	async login(_user: User): Promise<void> {
+		return;
+	}
+}
+
+export function createSAMLAuthRouter(providerName: string) {
+	const router = Router();
+
+	router.get(
+		'/metadata',
+		asyncHandler(async (req, res) => {
+			const { sp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			return res.header('Content-Type', 'text/xml').send(sp.getMetadata());
+		})
+	);
+
+	router.get(
+		'/',
+		asyncHandler(async (req, res) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { context: url } = await sp.createLoginRequest(idp, 'redirect');
+			const parsedUrl = new URL(url);
+			if (req.query.redirect) {
+				parsedUrl.searchParams.append('RelayState', req.query.redirect as string);
+			}
+			return res.redirect(parsedUrl.toString());
+		})
+	);
+
+	router.post(
+		'/logout',
+		asyncHandler(async (req, res) => {
+			const { sp, idp } = getAuthProvider(providerName) as SAMLAuthDriver;
+			const { context } = await sp.createLogoutRequest(idp, 'redirect', req.body);
+			if (req.cookies[env.REFRESH_TOKEN_COOKIE_NAME]) {
+				res.clearCookie(env.REFRESH_TOKEN_COOKIE_NAME, COOKIE_OPTIONS);
```

Comment:
Sorry to be that guy, but I also believe we need to invalidate the `refresh_token` in the DB, correct @rijkvanzanten? If so, it might be easier to just call the `logout` method in the AuthenticationService.